### ISSUE TITLE
feat: ref should be an array

### ISSUE THREAD

**Issue by bonustrack** [snapshot-labs/boost#26]
We should change `ref` to be an array, currently if we want define within a Boost strategy multiple requirement like doing a vote on Snapshot and do a retweet, we wouldn't be able to mark both the address and the Twitter username as id that have already claimed.

**Comment by Orland0x** [comment]
could we instead create a combined single ref by hashing the individual ones? 

```
like ref = keccak256(abi.encode(voteID, tweetID));
```

**Comment by bonustrack** [comment]
No because we need to check each of the ref, not all at once. Otherwise if someone have multiple tweetId but a single voteId, he will be able to claim multiple time.

**Comment by Orland0x** [comment]
okay, well we will need to redesign some part as you cannot use an array as a key for a mapping, so we will not be able to store who has claimed so far [here](https://github.com/snapshot-labs/boost/blob/04ad5dc8bc742b6e8324bdcee6190eee93c4b029/src/Boost.sol#L221)

ill have a think

**Comment by bonustrack** [comment]
@Orland0x Can't you just store each of the ref in the same variable `claimed`? And when someone claim you check all the refs against `claimed`?

**Comment by Orland0x** [comment]
> @Orland0x Can't you just store each of the ref in the same variable `claimed`? And when someone claim you check all the refs against `claimed`?

yeah we could do that, it adds 20k gas for each ref used though. 

> No because we need to check each of the ref, not all at once. Otherwise if someone have multiple tweetId but a single voteId, he will be able to claim multiple time.

The guard has to sign each claim so why cant they just check that refs have not been reused?

**Comment by bonustrack** [comment]
> The guard has to sign each claim so why cant they just check that refs have not been reused?

I'll need to think about it. If it's not enforced at the contract level the guard would need to assume everytime it issue a coupon that it can be claimed, but we may have user who request a coupon and not claim it directly then ask a coupon again. On solution could be to store the issued coupons so we can send it again
